### PR TITLE
Clarify some details about span kind and the meanings of the values.

### LINKED ISSUE #1360: The need for more granularity/clarity in CLIENT span conventions
**What are you trying to achieve?**

Currently, the span specification reads, when describing span kinds:

```
CLIENT Indicates that the span describes a synchronous request to some remote service. This span is the parent of a remote SERVER span and waits for its response.
```

This seems fine on the surface, but when you get into the details of actually writing instrumentation (and auto-instrumentation in particular), things get murky very quickly. Multiple layers of an application-framework stack can (and do) consider themselves logically as CLIENTs, even if none of them is actually opening up TCP connections and writing bytes to the wire.

For example, let's take a typical java HTTP call. Here are some possible spans that might want to be created by auto-instrumentation:

* JAX-RS HTTP client library makes a GET call
  * Apache HTTP client library makes a GET call
    * Java TCP stack opens up a socket
       * Java TCP stack requests that the OS makes a DNS lookup
       * Java TLS implementation negotiates TLS with the server 

(details here might be off, but I think the idea is clear).

So, which of these operations should create a CLIENT span? If library instrumentation for JAX-RS decides it should be a CLIENT, how can the Apache HTTP client library instrumentation create spans at all, given that a CLIENT has already been created (although it can't tell that, since the parent span is write-only, and the `kind` is immutable after the span has been created!).

At the moment, instrumentation does not have a way to deal with this situation, with the spec as-written. We should solve this problem and provide clarity on how instrumentation should be written in order to handle scenarios like this one.

I will also note, that it becomes even more confusing and difficult when dealing with databases that implement their protocol over HTTP (elastic, for example). Now we have database CLIENT spans that really need HTTP CLIENT spans below them, with all the additional complexity of observing the underlying network stack, as decsribed above.

### PR DESCRIPTION
**Pull request by jkwatson**
Fixes #1360

## Changes
Provides some additional clarity around Span.Kind and the meanings of the various values.

### CODE REVIEW

**Review by yurishkuro — APPROVED** [review]

**Review by carlosalberto — APPROVED** [review]

**Review by reyang — APPROVED** [review]

File: specification/trace/api.md [hunk 1]
```diff
@@ -697,7 +697,9 @@ and its children in a Trace.  `SpanKind` describes two independent
 properties that benefit tracing systems during analysis.
 
 The first property described by `SpanKind` reflects whether the Span
-is a remote child or parent.  Spans with a remote parent are
+is a "logical" remote child or parent. By "logical", we mean that
+the span is logically a remote child or parent, from the point of view
```
Comment (iNikem) [inline specification/trace/api.md:701]: In this section you say both "remote child" and "span with a remote parent". I am afraid "remote child" will be somewhat confusing term. Or at least we have to explicitly define what is "remote child".

**Review by iNikem — CHANGES_REQUESTED** [review]
I am not sure that these changes resolved #1360 completely.

Also, if the goal of these changes is to allow nested CLIENT spans, then I am against it :) For a very specific reason: if we have several CLIENT spans during one remote call, which of those spans should be propagated?

**Comment by jkwatson** [comment]
> I am not sure that these changes resolved #1360 completely.
> 
> Also, if the goal of these changes is to allow nested CLIENT spans, then I am against it :) For a very specific reason: if we have several CLIENT spans during one remote call, which of those spans should be propagated?

I think it's very simple: the last one to touch the headers is the one that is propagated. Is there more complication that I'm not considering?

**Comment by iNikem** [comment]
I removed my objection after offline discussion with @jkwatson

**Comment by trask** [comment]
I agree it's needed to relax this constraint on nesting. I'm wondering if there's anything we can do to make it easier on backends that use CLIENT/SERVER and PRODUCER/CONSUMER spans to draw service maps, e.g. we had this discussion in Java a while ago before we started suppressing nested CLIENT spans: https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/440

**Comment by carlosalberto** [comment]
Let's merge this after the 1.4 release as we ha enough approvals.

**Pull request merged**